### PR TITLE
Add Tier Three to smoke & cron e2e checkout tests

### PR DESCRIPTION
**Pull request by tjmw**
## What are you doing in this PR?

Add a smoke test (card) and a cron test (PayPal) for Tier Three checkout.

I've done a bit of refactoring to pull some user details fixtures out into a separate file so they can be re-used.

[**Trello Card**](https://trello.com/c/L8LD1pz3/1270-tier-3-e2e-test)

## Why are you doing this?

We'd like to know if Tier Three checkouts are broken in prod.

## How to test

Locally we can run:

`$ yarn test-smoke-ui`
`$ yarn test-cron-ui`

## Accessibility test checklist

-  [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-  [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-  [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-  [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: support-e2e/tests/test/checkout.ts [hunk 1]
```diff
@@ -1,17 +1,52 @@
...
+// shared types from the product catalog.
 type TestDetails = {
 	product: string;
 	ratePlan: string;
 	paymentType: string;
 	internationalisationId: string;
-	paymentFrequency: string;
+};
+
+const setUserDetailsForProduct = async (
```
Comment (tjmw) [inline support-e2e/tests/test/checkout.ts:22]: I introduced this method because we now need different user details depending on the product/country (i.e. Tier Three needs full address data). In future it might be nice to do this in a more generic way, e.g. with some kind of mapping, but I think this'll do for now.

File: support-e2e/package.json [hunk 2]
```diff
@@ -24,10 +24,12 @@
 	},
 	"devDependencies": {
 		"@guardian/prettier": "^8.0.0",
+		"@types/node": "^22.10.1",
 		"dotenv": "^16.3.1",
 		"husky": "^9.0.11",
 		"lint-staged": "^15.2.2",
-		"prettier": "^3.0.3"
+		"prettier": "^3.0.3",
+		"typescript": "^5.7.2"
```
Comment (tjmw) [inline support-e2e/package.json:32]: This has been added as there's a change in this PR to allow running `yarn tsc` in the support-e2e project to check the types. Playwright itself [doesn't check types](https://playwright.dev/docs/test-typescript) and I found myself wanting to do this. This could be pulled out to a separate PR if we preferred.

File: support-e2e/tests/test/checkout.ts [hunk 3]
```diff
@@ -1,17 +1,52 @@
...
+import { TestFields, ukWithPostalAddressOnly } from '../utils/userFields';
 
+// TODO: it'd be great to make the types here more specific, possibly using the
+// shared types from the product catalog.
 type TestDetails = {
 	product: string;
 	ratePlan: string;
 	paymentType: string;
 	internationalisationId: string;
-	paymentFrequency: string;
```
Comment (tjmw) [inline support-e2e/tests/test/checkout.ts:14]: This wasn't ever used.

File: support-e2e/tests/tieredCheckout.test.ts [hunk 4]
```diff
@@ -10,166 +10,94 @@ import {
...
+} from './utils/userFields';
 
 export interface TestDetails {
 	tier: 1 | 2 | 3;
-	ratePlan: 'Monthly' | 'Annual';
+	ratePlan: 'Monthly' | 'Annual' | 'DomesticMonthly';
 	paymentType: 'Credit/Debit card' | 'Direct debit' | 'PayPal';
 	fields: TestFields;
-	internationalisationId?: 'US' | 'AU';
+	internationalisationId: 'UK' | 'US' | 'AU' | 'EU';
```
Comment (tjmw) [inline support-e2e/tests/tieredCheckout.test.ts:29]: Force us to be explicit about this instead of defaulting `undefined` to `UK`.

File: support-e2e/tests/utils/testUserDetails.ts [hunk 5]
```diff
@@ -18,31 +19,30 @@ export const setTestUserRequiredDetails = async (
 
 export const setTestUserDetails = async (
 	page: Page,
-	testDetails: TestDetails,
```
Comment (tjmw) [inline support-e2e/tests/utils/testUserDetails.ts:21]: There are fields in `TestDetails` which this function doesn't care about, so I've changed to interface. This allows us to use this method elsewhere, only providing the data needed.

File: support-e2e/tests/tieredCheckout.test.ts [hunk 6]
```diff
@@ -10,166 +10,87 @@ import {
...
 }
 
-const testsDetails: TestDetails[] = [
+export const testsDetails: TestDetails[] = [
 	{
 		tier: 1,
 		ratePlan: 'Monthly',
 		paymentType: 'Direct debit',
-		fields: { email: email(), firstName: firstName(), lastName: lastName() },
+		fields: personalDetailsOnly(),
```
Comment (paul-daniel-dempsey) [inline support-e2e/tests/tieredCheckout.test.ts:37]: Nice & neat moving this out

File: support-e2e/tests/utils/testUserDetails.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import { Page } from '@playwright/test';
 import { TestDetails } from '../tieredCheckout.test';
```
Comment (paul-daniel-dempsey) [inline support-e2e/tests/utils/testUserDetails.ts:2]: probably don;t need this now

File: support-e2e/tests/test/checkout.ts [hunk 8]
```diff
@@ -1,17 +1,52 @@
...
+};
+
+const setUserDetailsForProduct = async (
+	page,
+	product,
+	internationalisationId,
+) => {
+	switch (product) {
+		case 'SupporterPlus':
```
Comment (paul-daniel-dempsey) [inline support-e2e/tests/test/checkout.ts:28]: In a seperate PR, maybe we could rename to `All-Access Digital` and maybe introduce a Tier1 `Support` test?

File: support-e2e/tests/utils/testUserDetails.ts (hunk 5 above)
Comment (paul-daniel-dempsey) [reply support-e2e/tests/utils/testUserDetails.ts:21]: Don't see any optional parameters but probably saves having to build that whole confusingly named `TestDetails` object? Plus more concise use of (shorter)  field names within this function is a bonus.

File: support-e2e/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (paul-daniel-dempsey) [inline support-e2e/tsconfig.json:1]: Presume this goes with the type checking?

File: support-e2e/package.json (hunk 2 above)
Comment (paul-daniel-dempsey) [reply support-e2e/package.json:32]: If stays maybe a note in main conversation window on how to use?

**Review by paul-daniel-dempsey — APPROVED** [review]
Looks Greate, couple of (my) notes really, maybe add something on the type checking on PR? Cron and Smoke tests pass perfectly.

File: support-e2e/tsconfig.json (hunk 9 above)
Comment (tjmw) [reply support-e2e/tsconfig.json:1]: Yeah exactly, I needed a basic tsconfig to do this 👍

File: support-e2e/tests/utils/testUserDetails.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tjmw) [reply support-e2e/tests/utils/testUserDetails.ts]: Good spot! Removed in 2eea8f474c76cb6237925863d7abfbcadd896ea6.

File: support-e2e/package.json (hunk 2 above)
Comment (tjmw) [reply support-e2e/package.json:32]: I've added a note to the README about being able to run `yarn tsc` to type check the tests.
